elastic/apm-server
# Flush aggregator on shutdown

## Motivation/summary

Flush aggregated transaction metrics on shutdown, to minimise data loss. This could be important in a rolling upgrade.

There are some significant, and necessary, changes to the publisher to support an interruptible graceful shutdown-with-timeout. The publisher's Stop method now takes a context, and waits until the context is cancelled, or the published events are acknowledged, before returning. This is all necessary because the shutdown timeout, from the user's point of view, should encompass the entire shutdown procedure -- not just closing the pipeline client.

## Checklist


I have considered changes for:
~- [ ] documentation~
~- [ ] logging (add log lines, choose appropriate log selector, etc.)~
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~
~- [ ] telemetry~
~- [ ] Elasticsearch Service (https://cloud.elastic.co)~
~- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)~
~- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)~

## How to test these changes

1. Enable transaction aggregation
2. Send some transactions
3. Shutdown APM Server (Ctrl+C) before the aggregation interval elapses
4. Check that metrics are published within the shutdown interval (default 5s)

## Related issues

Closes #3789
